### PR TITLE
fix(Treeview): throwing NRE on Dragging

### LINKED ISSUE #16041: [TreeView][Skia.Gtk] TreeView throws NullReferenceException when dragging parent onto descendant
### Current behavior

When dragging an item onto one of its descendants.

image

### Expected behavior

This error doesn't occur on Windows. Looks likes Windows collapses an element as it starts to drag thereby preventing this situation in the first place.

### How to reproduce it (as minimally and precisely as possible)

[TreeViewDragRepro.zip](https://github.com/unoplatform/uno/files/14773748/TreeViewDragRepro.zip)

https://playground.platform.uno/#3a138698 (doesn't reproduce)

### Workaround

_No response_

### Works on UWP/WinUI

Yes

### Environment

Uno.WinUI / Uno.WinUI.WebAssembly / Uno.WinUI.Skia

### NuGet package version(s)

5.1.104

### Affected platforms

Skia (GTK on Linux/macOS/Windows)

### IDE

Visual Studio 2022

### IDE version

_No response_

### Relevant plugins

_No response_

### Anything else we need to know?

_No response_

### PR DESCRIPTION
**Pull request by ramezgerges**
GitHub Issue (If applicable): closes #16041

## PR Type

What kind of change does this PR introduce?

## PR Checklist

Please check if your PR fulfills the following requirements:

- [ ] Docs have been added/updated which fit [documentation template](https://github.com/unoplatform/uno/blob/master/doc/.feature-template.md) (for bug fixes / features)
- [ ] [Unit Tests and/or UI Tests](https://github.com/unoplatform/uno/blob/master/doc/articles/uno-development/working-with-the-samples-apps.md) for the changes have been added (for bug fixes / features) (if applicable)
- [ ] Validated PR `Screenshots Compare Test Run` results.
- [ ] Contains **NO** breaking changes
- [ ] Associated with an issue (GitHub or internal) and uses the [automatic close keywords](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue).
- [ ] Commits must be following the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/#summary) specification.

## Other information

Internal Issue (If applicable):

### CODE REVIEW

File: src/Uno.UI.RuntimeTests/Tests/Microsoft_UI_Xaml_Controls/Given_TreeView.cs [hunk 1]
```diff
@@ -94,6 +95,101 @@ public async Task When_TreeViewItem_Dragged_Near_the_Edge()
...
+
+			// The bug is extremely hard to reproduce. Even slight changes to the numbers will fail to repro.
+			// The main idea is to have the first PointerEnteredEvent in the "Parent" TreeViewItem
+			// and the very next "frame" (i.e. pointer event) should jump to the the "Child" TreeViewItem.
+			// At least, that's the leading hypothesis.
+			// Note that removing the WaitForIdle's will throw an NRE and this is somewhat expected. The TreeView
+			// removes the children "nodes" immediately when a parent node starts dragging, but the children
+			// are asynchronously removed, so if an "about to be removed" child node receives a DragOver/DragEnter,
+			// the event callback will run some logic that assumes that the node is still in the TreeView, when it
+			// technically isn't (it's in the tree, but the TreeView has internally discarded it), which throws the NRE.
```
Comment (agneszitte) [inline src/Uno.UI.RuntimeTests/Tests/Microsoft_UI_Xaml_Controls/Given_TreeView.cs:179]:
```suggestion
			// The bug is extremely hard to reproduce. Even slight changes to the numbers will fail to repro.
			// The main idea is to have the first PointerEnteredEvent in the "Parent" TreeViewItem
			// and the very next "frame" (i.e. pointer event) should jump to the "Child" TreeViewItem.
			// At least, that's the leading hypothesis.
			// Note that removing the WaitForIdle(s) will throw an NRE and this is somewhat expected. The TreeView
			// removes the children "nodes" immediately when a parent node starts dragging, but the children
			// are asynchronously removed, so if an "about to be removed" child node receives a DragOver/DragEnter,
			// the event callback will run some logic that assumes that the node is still in the TreeView, when it
			// technically isn't (it's in the tree, but the TreeView has internally discarded it), which throws the NRE.
```

**Review by jeromelaban — APPROVED** [review]

File: src/Uno.UI.RuntimeTests/Tests/Microsoft_UI_Xaml_Controls/Given_TreeView.cs [hunk 2]
```diff
@@ -94,6 +95,104 @@ public async Task When_TreeViewItem_Dragged_Near_the_Edge()
 		}
 #endif
 
+		[TestMethod]
+#if __ANDROID__ || __IOS__
+		[Ignore("Fails")]
```
Comment (Xiaoy312) [inline src/Uno.UI.RuntimeTests/Tests/Microsoft_UI_Xaml_Controls/Given_TreeView.cs:100]: any explanation here?

File: src/Uno.UI.RuntimeTests/Tests/Microsoft_UI_Xaml_Controls/Given_TreeView.cs [hunk 3]
```diff
@@ -94,6 +95,104 @@ public async Task When_TreeViewItem_Dragged_Near_the_Edge()
...
+			treeView.RootNodes[0].IsExpanded = true;
+			await WindowHelper.WaitForIdle();
+
+			Assert.AreEqual(2, treeView.FindVisualChildByType<ItemsStackPanel>().Children.Count);
+		}
+
+#if HAS_UNO
+		// https://github.com/unoplatform/uno/issues/16041
+		// This test actually always passes due to catching the NRE in SafeRaiseEvent in RaiseDragEnterOrOver
+		// so you need to check manually if there are any logged exceptions
```
Comment (Xiaoy312) [inline src/Uno.UI.RuntimeTests/Tests/Microsoft_UI_Xaml_Controls/Given_TreeView.cs:142]: i think you can actually catch the exception to make the test pass/fail accordingly:
```cs
partial class TestableTreeView : TreeView
	public Action<DragEventArgs, Exception>? OnDragEnterExceptionHandler { get; set; }
	
	protected override void OnDragEnter(Microsoft.UI.Xaml.DragEventArgs args)
		try { base.OnDragEnter(args); }
		catch (Exception e) { OnDragEnterExceptionHandler?.Invoke(args, e); }
```

File: src/Uno.UI.RuntimeTests/Tests/Windows_UI_Xaml_Controls/Given_ListViewBase.cs [hunk 4]
```diff
@@ -3972,6 +3972,38 @@ public async Task When_ItemsSource_NotNotify_Changed(string sutType, string scen
 #endif
 		}
 
+		[TestMethod]
+#if __ANDROID__ || __IOS__
+		[Ignore("Fails")]
```
Comment (Xiaoy312) [inline src/Uno.UI.RuntimeTests/Tests/Windows_UI_Xaml_Controls/Given_ListViewBase.cs:3977]: same here, an explanation would be nice

**Review by Xiaoy312 — APPROVED** [review]

File: src/Uno.UI.RuntimeTests/Tests/Microsoft_UI_Xaml_Controls/Given_TreeView.cs (hunk 2 above)
Comment (ramezgerges) [reply src/Uno.UI.RuntimeTests/Tests/Microsoft_UI_Xaml_Controls/Given_TreeView.cs:100]: The changes in this PR fix this only for the managed VirtualizingPanels. I'm not sure what the situation with mobile targets is, but a corresponding fix will need to be added there as well. I'm not very familiar with the virtualization implementation on iOS and Android (and also I can't test locally), so I was hoping to get some help there.

Comment (Xiaoy312) [reply src/Uno.UI.RuntimeTests/Tests/Microsoft_UI_Xaml_Controls/Given_TreeView.cs:100]: you could add this in a comment

**Pull request merged**